File: leet-code/78. Subsets/Subsets78_2.java
package com.example;

import java.util.List;
import java.util.ArrayList;

public class Subsets78_2 {
    public List<List<Integer>> subsets(int[] nums) {
        return getSubsets(new ArrayList<>(), new ArrayList<>(), nums, 0);
    }

    private List<List<Integer>> getSubsets(List<List<Integer>> answer, List<Integer> subset, int[] nums, int idx) {
        answer.add(List.copyOf(subset));
        for (int i = idx; i < nums.length; i++) {
            subset.add(nums[i]);
            getSubsets(answer, subset, nums, i + 1);
            subset.remove(subset.size() - 1);
        }
        return answer;
    }
}

File: leet-code/78. Subsets/Subsets78_2.kt
package com.example.algorithm

class Subsets78_2 {
    fun subsets(nums: IntArray): List<List<Int>> {
        return getSubsets(nums, 0, mutableListOf(), mutableListOf())
    }

    private fun getSubsets(nums: IntArray, idx: Int, subsetList: MutableList<List<Int>>, subset: MutableList<Int>): List<List<Int>> {
        subsetList.add(subset.toList())
        if (nums.size == idx) {
            return subsetList
        }
        for (i in idx until nums.size) {
            subset.add(nums[i])
            getSubsets(nums, i + 1, subsetList, subset)
            subset.removeLast()
        }
        return subsetList
    }
}
